[chore][tracker]: save most recent (archive) write index to disk

VihasMakwana opened the pull request:
This PR stores the most recent index to disk. Much similar to what happens for persistent queue. It also adds `Batch` methods to `operator.Persister`, as saving the metadata and saving the index should be a transaction and it can only be achieved via `Batch`.

For eg. if user has configured archiving to store 100 poll cycles, let's assume:
- For first collector run, it stores 10 cycles and `archiveIndex` is 11 (pointing to the next index).
- When the collector is restarted, we will restore the `archiveIndex` from disk and continue from index 11

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32727

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Added UT for checking index is in-bounds and to test archive restoration.

VihasMakwana committed:
chore: save most recent written index

VihasMakwana commented on pkg/stanza/fileconsumer/internal/checkpoint/checkpoint.go:
```diff
@@ -37,8 +38,8 @@ func SaveKey(ctx context.Context, persister operator.Persister, rmds []*reader.M
-
-	if err := persister.Set(ctx, key, buf.Bytes()); err != nil {
+	ops = append(ops, storage.SetOperation(key, buf.Bytes()))
+	if err := persister.Batch(ctx, ops...); err != nil {
```
For existing usage, this will be a no-op.

VihasMakwana committed:
chore: readability

VihasMakwana committed:
retunr

VihasMakwana committed:
test

VihasMakwana committed:
lint

VihasMakwana committed:
lint and test

VihasMakwana committed:
gci

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -52,13 +55,18 @@ type fileTracker struct {
 
+var errInvalidValue = errors.New("invalid value")
+
+var archiveIndexKey = "knonwFiles_ai"
```
Let's be verbose so the valu is self explanatory

```suggestion
var archiveIndexKey = "knownFilesArchiveIndex"
```

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -144,6 +155,26 @@ func (t *fileTracker) TotalReaders() int {
 }
 
+func (t *fileTracker) restoreArchiveIndex() {
+	if !t.archiveEnabled() {
+		return
+	}
```
nit: let's check this before calling the function. Maybe just my preference but it makes more sense to me to say "if it's enable, do it" rather than "do it. is it enabled?"

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -144,6 +155,26 @@ func (t *fileTracker) TotalReaders() int {
+		// It's best if we reset the index or else we might end up writing invalid keys
+		t.set.Logger.Warn("the read index was found, but it exceeds the bounds. Starting from 0")
+		t.archiveIndex = 0
+	}
```
Good idea to check for this case.

However, I wonder if we can handle it better than restarting from zero. What would it take to search the archive for the most recently updated?

I think we could maintain some kind of data structure which notes the time each archive was written. Maybe just `map[index]time.Time`. Then when we first create the tracker, we can load this up and find the most recent timestamp. We can also check for the case where `pollsToArchive` has changed and then rewrite the storage to align with the new value. 

For example, if we previously saved 10 archives and find that `pollsToArchive` is now 5, we can find the 5 most recent indices based on the timestamp structure, then rewrite the archive files so that these are 0-4. We should probably even delete the extras from storage as well.

VihasMakwana replied:
This solution does makes sense to me, but it becomes tricky when we eventually overwrite old archive data, as it is a ring buffer.
We might need to load the filesets in memory. 
I'll find a few ways.

djaglowski replied:
> it becomes tricky when we eventually overwrite old archive data, as it is a ring buffer.

Can you elaborate?

> We might need to load the filesets in memory.

If it's more than one at a time then it defeats the point of the archive.

VihasMakwana replied:
> Can you elaborate?

Consider this archive,

<img width="511" alt="Screenshot 2024-12-23 at 8 37 38 PM" src="https://github.com/user-attachments/assets/4e5d8bec-fec1-4422-89bf-81ac1827853d" />

We've rolled over once and the latest data is at index `4` and `archiveIndex` (i.e. where the next data will be written) is at index `5`. 

Let's suppose that new `polls_to_archive` is 7.
We now need to construct a new, smaller archive with 7 most recent elements. 
These elements are (from most recent to least recent):

```14, 13, 12, 11, 10, 9, 8```

We cannot simply rewrite archive in-place without caching values. 

It would be much simpler to convert archive like following image,

<img width="572" alt="Screenshot 2024-12-23 at 8 41 43 PM" src="https://github.com/user-attachments/assets/e5492c97-992c-491d-b511-16b46fbd5aa1" />

and we would delete excess data. 

Wdyt?

VihasMakwana replied:
> What would it take to search the archive for the most recently updated?

It would always be data stored at `archiveIndex-1` index. We will store `archiveIndex` on disk, so in next collector run, we would load that value and we can find most recent data.

`archiveIndex` points at the next location where data will be written. 
This can point to either of following:
- Least recent data
- Pointing to an empty slot (archive is partially filled)

djaglowski replied:
> It would always be data stored at archiveIndex-1 index.

What if `pollsToArchive < archiveIndex-1`?

Anyways, this is an edge case we can worry about later. I think we have bigger problems to work through first.

VihasMakwana replied:
Correcting my previous comment, 
Most recently written index is `(archiveIndex-1) % previouspollsToArchive `

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -162,13 +193,15 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
 	//                   index
 
-	if t.pollsToArchive <= 0 || t.persister == nil {
+	if !t.archiveEnabled() {
```
Let's move this check out too

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -295,3 +332,18 @@ func (t *noStateTracker) EndPoll() {}
 
 func (t *noStateTracker) FindFiles([]*fingerprint.Fingerprint) []*reader.Metadata { return nil }
+
+func intToByte(val int) []byte {
+	return binary.LittleEndian.AppendUint64([]byte{}, uint64(val))
+}
+
+func byteToIndex(buf []byte) (int, error) {
+	if buf == nil {
+		return 0, nil
+	}
+	// The sizeof uint64 in binary is 8.
+	if len(buf) < 8 {
+		return 0, errInvalidValue
+	}
+	return int(binary.LittleEndian.Uint64(buf)), nil
+}
```
Can we reuse the same marshal/unmarshal strategy from `checkpoint`? https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b6238cfe3c1988309cab87056a4bedcc0f905552/pkg/stanza/fileconsumer/internal/checkpoint/checkpoint.go#L25-L41

VihasMakwana committed:
Update pkg/stanza/fileconsumer/internal/tracker/tracker.go

Co-authored-by: Daniel Jaglowski <jaglows3@gmail.com>

VihasMakwana committed:
chore: archive restoration

VihasMakwana committed:
chore: add docs

VihasMakwana committed:
linting

VihasMakwana commented:
@djaglowski I've added documentation and implemented the archive restoration.

I think adding a new data structure like `map[index]time.Time` will add to the complexity if we've rolled over and overwriting older data. 

We can accomplish archive restoration without any new data structure and I've added a document to highlight this. Please take a look and let me know your thoughts.

VihasMakwana committed:
Merge branch 'main' into archive-minor-fix

djaglowski commented on pkg/stanza/fileconsumer/design/design.md:
```diff
@@ -0,0 +1,279 @@
+# File Matching
```
How much of this is duplicated from https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/operator/input/file/design.md?

djaglowski commented on pkg/stanza/fileconsumer/design/archive.md:
```diff
@@ -0,0 +1,173 @@
+# File archiving
+
+The file consumer now supports archiving. Previously, file offsets older than three poll cycles were discarded, and if such files reappeared (which could happen if they were temporarily removed or if `exclude_older_than` was enabled), the entire file contents would be read again.
```
I don't think we need to describe previous functionality in this document

djaglowski commented on pkg/stanza/fileconsumer/design/archive.md:
```diff
@@ -0,0 +1,173 @@
+
+## How does archiving work?
+
+- We stores the offsets older than three poll cycles on disk. If we use `polls_to_archive: 10`, the on-disk structure looks like following:
```
Do we have 3 in memory and 7 on disk? This seems worth calling out explicitly in the example.

VihasMakwana replied:
> Do we have 3 in memory and 7 on disk? This seems worth calling out explicitly in the example.

No. We have 3 in memory and 10 on disk. 
I will explicitly mention this to clear the difference.

djaglowski commented on pkg/stanza/fileconsumer/design/archive.md:
```diff
@@ -0,0 +1,173 @@
+### How does reading from archiving work?
+
+During reader creation, we group all the new (or unmatched) files and try to find a match in archive. From high level, it consists of following steps:
+1. We start from most recently writen index on archive and load the data from it.
```
We should mention in-memory first, then archive is used as a fallback.

VihasMakwana replied:
Got it.

djaglowski reviewed (commented):
Can we move the documentation changes to another PR?

VihasMakwana committed:
doc update

VihasMakwana committed:
remove doc

VihasMakwana committed:
remove image

VihasMakwana commented:
@djaglowski as we'll include docs in new PR, please let me know if you have any comments regarding the implementation.

VihasMakwana committed:
chore: ci bug

VihasMakwana committed:
Merge branch 'main' into archive-minor-fix

VihasMakwana commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -144,6 +161,109 @@ func (t *fileTracker) TotalReaders() int {
+		previousPollsToArchive = t.pollsToArchive
+	}
+
+	t.archiveIndex, err = decodeIndex(byteIndex)
+	if err != nil {
+		t.set.Logger.Error("error getting read index. Starting from 0", zap.Error(err))
```
> Does it need to be tied into a conditional with pollsToArchive

not really. we can move it at the beginning of the function.

VihasMakwana commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -144,6 +161,109 @@ func (t *fileTracker) TotalReaders() int {
+}
+
+func (t *fileTracker) rewriteArchive(ctx context.Context, previousPollsToArchive int) {
+	// Ensure archiveIndex is non-negative
```
We can remove this check. It shouldn't happen.
If there's any error, we'll notice it while reading or converting bytes to int.

VihasMakwana commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -144,6 +161,109 @@ func (t *fileTracker) TotalReaders() int {
+		return
+	}
+	// Function to swap data between two indices
+	swapData := func(idx1, idx2 int) error {
```
Yeah. I just realized that we don't need to "swap". 
We just need to read from old "key" and rewrite it to new "key".

VihasMakwana replied:
@djaglowski 

> Isn't this terribly expensive? Why do we need it?

I just realised I forgot to answer this aspect of your question. As per [our discussion](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36799#discussion_r1884274441), we might need to rewrite the archive if `pollsToArchive` has changed. 
For eg. if we previously saved 10 archives and find that pollsToArchive is now 5, we can find the 5 most recent indices based on the timestamp structure, then rewrite the archive files so that these are 0-4

For rewriting, we need to perform two operations i.e. reading from old key and writing to new one.

That being said, this is a one time operation and this will only happen if `pollsToArchive` has decreased. We will not attempt to rewrite if archive size is same or if it has increased.

Please let me know if you have any other questions.

> Isn't this terribly expensive

Yes, my initial approach to "swap" was terribly expensive. I've removed the unnecessary swapping and optimised it to cater our need to rewrite.

VihasMakwana committed:
chore: test cases, ctx

VihasMakwana committed:
Merge branch 'main' into archive-minor-fix

VihasMakwana committed:
update go.mod

VihasMakwana commented:
@djaglowski I've added more test coverage. PTAL at your convenience! Thank you.

VihasMakwana committed:
ci

VihasMakwana committed:
lint

VihasMakwana committed:
Merge branch 'main' into archive-minor-fix

djaglowski committed:
Merge branch 'main' into archive-minor-fix
